File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> pq;
        
        int k=lists.size();
        for(int i=0; i<k; i++){
            if(lists[i] != NULL){
                pq.push({lists[i]->val, lists[i]});
            }
        }
        
        if(pq.size()==0) return NULL;
        
        ListNode* dh = new ListNode(-1);
        ListNode* dt = dh;
        
        while(pq.size()>0){
            
            pair<int, ListNode*> curr = pq.top();
            pq.pop();
            
            int val = curr.first;
            ListNode* add = curr.second;
            
            dt->next = add;
            dt = dt->next;
            
            if(add->next != NULL){
                pq.push({add->next->val, add->next});
            }
        }
        
        return dh->next;
    }
};

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge(self, l1, l2):
        
        if l1 == None or l2 == None:
            if l1 != None: return l1
            else: return l2
        
        c1, c2 = l1, l2
        
        dh = ListNode(-1)
        dt = dh
        
        while c1 and c2:
            if c1.val < c2.val:
                dt.next = c1
                c1 = c1.next
            else:
                dt.next = c2
                c2 = c2.next
            dt = dt.next
        
        if c1:
            dt.next = c1
        if c2:
            dt.next = c2
        
        return dh.next
    
    def mergeSort(self, lists, si, ei):
        if si > ei:
            return None
        if si == ei:
            return lists[si]
        
        mid = (si + ei)//2
        
        l1 = self.mergeSort(lists, si, mid)
        l2 = self.mergeSort(lists, mid+1, ei)
        
        return self.merge(l1, l2)
    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        if len(lists) == 0:
            return None
        
        return self.mergeSort(lists, 0, len(lists)-1)